### ISSUE TITLE
关于预测的bev结果可视化

### ISSUE THREAD

**Issue by qwefvb** [LogosRoboticsGroup/SeerDrive#5]
作者您好，我使用了您的开源权重进行了推理，并可视化了bev_map, fut_bev_map。但是它们似乎并没有学到什么东西。
我对预测结果bev_semantic_map和fut_bev_semantic_map采用了如下两套逻辑的可视化（这两套逻辑我在diffusiondrive上都可以正常显示bev map结果），但是结果很差，几乎全是一个类别，且得分很低。

我的可视化逻辑如下

<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/dbd1fa1d-eb5f-4122-8f85-7d8e9da3ebc2" />
<img width="1200" height="700" alt="Image" src="https://github.com/user-attachments/assets/ea479b6b-d03a-4599-8e83-0b45e6f12ff0" />
<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/f62a91e0-cc9e-4805-9480-08325ccdb66c" />
<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/ed0199b7-5672-45cf-a3cb-4012658217ff" />
<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/75ce217c-5b4e-482e-b661-464f23634efb" />
<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/a14628a6-92da-4c18-8719-9cfacc1e888d" />
<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/49b5c0de-f81f-4e78-a8eb-1dd511a18f79" />
<img width="500" height="500" alt="Image" src="https://github.com/user-attachments/assets/bf5ee0c4-4092-46f1-a7cb-5b936759ae88" />

可视化逻辑1:
def add_argmax_bev_map_to_ax(
        ax: plt.Axes,
        fut_bev_map: Union[np.ndarray, torch.Tensor],
        extent: Optional[List[float]] = None
) -> plt.Axes:

    # 1. Convert to Numpy
    if hasattr(fut_bev_map, 'detach'):
        fut_bev_map = fut_bev_map.detach().cpu().numpy()

    # 2. Handle Dimensions
    if fut_bev_map.ndim == 4:
        fut_bev_map = fut_bev_map[0]

    c, h, w = fut_bev_map.shape  # C=8, H=128, W=256

    # 3. Apply Argmax to get the dense label map
    # Result: [H, W] array of integer labels (0-7)
    labels_map = np.argmax(fut_bev_map, axis=0).astype(np.int32)

    # 4. Prepare Color Mapping Array (for efficient coloring)
    # The array is [8, 4] for 8 classes (0-7) in RGBA format
    color_map_array = np.zeros((c, 4), dtype=np.float32)

    # Map colors for indices 1 through 7
    for key in BEV_SEMANTIC_CLASSES.keys():
        if key >= c: continue

        # 使用 get_color_and_config 获取颜色。由于是硬分割，我们使用全不透明度 1.0
        color_hex, _, _ = get_color_and_config(key)
        rgba = to_rgba(color_hex, alpha=1.0)
        color_map_array[key] = rgba

    # Index 0 (Background) remains (0, 0, 0, 0), i.e., fully transparent.

    # 5. Efficiently color the image using NumPy indexing
    # final_image[h, w] = color_map_array[labels_map[h, w]]
    final_image = color_map_array[labels_map]  # Result: [H, W, 4] RGBA image

    # 6. Calculate Extent (if not provided, same logic as before)
    if extent is None:
        cfg = SeerDriveConfig()
        y_span = cfg.bev_pixel_width * cfg.bev_pixel_size
        x_span = cfg.bev_pixel_height * cfg.bev_pixel_size
        extent = [-y_span / 2, y_span / 2, -x_span / 2, x_span / 2]

    # Argmax Visualization has a single zorder
    zorder_val = 5

    # 7. Plotting
    ax.imshow(
        final_image,
        extent=extent,
        origin='lower',
        zorder=zorder_val,
        interpolation='nearest'
    )
    ax.invert_xaxis()

    return ax

可视化逻辑2:
def softmax_helper(x: np.ndarray, axis: int = 0) -> np.ndarray:
    """Computes softmax over a specific axis for numpy arrays."""
    # Subtract max for numerical stability
    e_x = np.exp(x - np.max(x, axis=axis, keepdims=True))
    return e_x / e_x.sum(axis=axis, keepdims=True)

def add_fut_bev_map_to_ax(
        ax: plt.Axes,
        fut_bev_map,
        threshold: float = 0.3,  # Lower threshold since probabilities are distributed across 8 classes
        extent: Optional[List[float]] = None,
        is_logits: bool = True
) -> plt.Axes:

    # 1. Convert to Numpy
    if hasattr(fut_bev_map, 'detach'):
        # It's a torch tensor
        fut_bev_map = fut_bev_map.detach().cpu().numpy()

    # 2. Handle Dimensions
    # Remove batch dimension if present: [1, 8, 128, 256] -> [8, 128, 256]
    if fut_bev_map.ndim == 4:
        fut_bev_map = fut_bev_map[0]

    # 3. Apply Softmax (Logits -> Probabilities)
    if is_logits:
        # We apply softmax across channel dimension (axis 0)
        # This converts raw scores into probabilities summing to 1.0 per pixel
        fut_bev_map = softmax_helper(fut_bev_map, axis=0)

    c, h, w = fut_bev_map.shape

    # 5. Plotting Loop
    # Sort by z-order to draw roads first, then vehicles
    sorted_keys = sorted(BEV_SEMANTIC_CLASSES.keys(), key=lambda k: get_color_and_config(k)[2])

    for key in sorted_keys:
        if key >= c: continue

        # Get probability map for this class
        heatmap = fut_bev_map[key]

        # Thresholding: Only draw pixels where this class probability is high enough
        mask = heatmap > threshold

        if not np.any(mask):
            continue

        color_hex, alpha_val, zorder_val = get_color_and_config(key)
        r, g, b, _ = to_rgba(color_hex)

        # Create RGBA layer
        rgba_img = np.zeros((h, w, 4))
        rgba_img[mask, 0] = r
        rgba_img[mask, 1] = g
        rgba_img[mask, 2] = b

        # Optional: Scale alpha by probability confidence?
        # For now, we use constant alpha for visibility
        rgba_img[mask, 3] = alpha_val

        ax.imshow(
            rgba_img,
            extent=extent,
            origin='lower',
            zorder=zorder_val,
            interpolation='nearest'
        )
    ax.invert_xaxis()
    return ax

**Comment by Zbozhou** [comment]
我不太清楚您的这个脚本可能存在什么问题，不过可视化逻辑不用这么复杂，labels_map = np.argmax(fut_bev_map, axis=0)到这一步之后，直接把0-7的各个类别用不同颜色表达出来就行

**Comment by qwefvb** [comment]
@Zbozhou 我确实是这么做的，但是结果存在问题，模型貌似并没有学到东西

**Comment by Zbozhou** [comment]
稍候，我试一下

**Comment by Zbozhou** [comment]
您好，我用如下的可视化方法，得到的的结果大概是下图所示，如果您觉得预测性能不够好，可以换为文章中的v2-99作为backbone，增大了backbone的预测效果更好。
```
pred_class = np.argmax(fut_bev, axis=0)  # [128, 256] 取每像素得分最大的位置（即预测的类别）
pred_class = np.fliplr(pred_class)  # 横轴镜像

# 定义颜色映射（8 个类别）
cmap = ListedColormap([
      "#d3d3d3",  # class 0: 黑色
      "#1f77b4",  # class 1: 蓝色
      "#2ca02c",  # class 2: 绿色
      "#d62728",  # class 3: 红色
      "#9467bd",  # class 4: 紫色
      "#ff7f0e",  # class 5: 橙色
      "#8c564b",  # class 6: 棕色
      "#e377c2",  # class 7: 粉红色
])
plt.figure(figsize=(10, 10))
plt.imshow(pred_class, cmap=cmap, origin='lower')
plt.tight_layout()
plt.savefig(os.path.join(save_dir, "future_bev_map.png"))
plt.close()
```

<img width="986" height="524" alt="Image" src="https://github.com/user-attachments/assets/59d59157-24b2-48fb-8120-8cf0ed239584" />

**Comment by qwefvb** [comment]
@Zbozhou 感谢你的答复，我已经找到我的问题，在SeerDrive_agent中，我并没有正确导入训练好的模型，我添加了下面的函数得到了解决：
    def init_from_pretrained(self):
        # import ipdb; ipdb.set_trace()
        if self._checkpoint_path:
            if torch.cuda.is_available():
                checkpoint = torch.load(self._checkpoint_path)
            else:
                checkpoint = torch.load(self._checkpoint_path, map_location=torch.device('cpu'))
    
            state_dict = checkpoint['state_dict']
    
            # Remove 'agent.' prefix from keys if present
            state_dict = {k.replace('agent.', ''): v for k, v in state_dict.items()}
    
            # Load state dict and get info about missing and unexpected keys
            missing_keys, unexpected_keys = self.load_state_dict(state_dict, strict=False)
    
            if missing_keys:
                print(f"Missing keys when loading pretrained weights: {missing_keys}")
            if unexpected_keys:
                print(f"Unexpected keys when loading pretrained weights: {unexpected_keys}")
        else:
            print("No checkpoint path provided. Initializing from scratch.")
